File: courses/codes/HeapExample.java
import java.util.List;
import java.util.ArrayList;

public class HeapExample {

    static List<Integer> H;

    public static void add(int x) {
        H.add(x);
        int idx = H.size() - 1;
        upheap(idx);
    }

    public static int peek() {
        return H.get(0);
    }

    public static int extractMin() {
        int minValue = peek();
        int last = H.get(H.size()-1);
        H.set(0, last);
        H.remove(H.size()-1);
        downheap(0);
        return minValue;
    }

    public static void upheap(int i) {
        int p = (i - 1) / 2;
        if (p == -1) return;
        if (H.get(p) > H.get(i)) {
            int temp = H.get(p);
            H.set(p, H.get(i));
            H.set(i, temp);
            upheap(p);
        }
    }

    public static void downheap(int i) {
        
        int lc = 2*i + 1;
        int rc = 2*i + 2;

        if (lc >= H.size()) return;
        else if (rc >= H.size()) {
            int temp = H.get(i);
            H.set(i, H.get(lc));
            H.set(lc, temp);
        } else {
            if (H.get(lc) <= H.get(rc)) {
                int temp = H.get(i);
                H.set(i, H.get(lc));
                H.set(lc, temp);
                downheap(lc);
            } else {
                int temp = H.get(i);
                H.set(i, H.get(rc));
                H.set(rc, temp);
                downheap(rc);
            }
        }
    }

    public static void main(String[] args) {
        H = new ArrayList<>();

        add(6);
        System.out.println("Min element: " + peek() + "; H = " + H);

        add(50);
        System.out.println("Min element: " + peek() + "; H = " + H);

        add(11);
        System.out.println("Min element: " + peek() + "; H = " + H);

        add(25);
        System.out.println("Min element: " + peek() + "; H = " + H);

        add(42);
        System.out.println("Min element: " + peek() + "; H = " + H);

        add(20);
        System.out.println("Min element: " + peek() + "; H = " + H);

        add(98);
        System.out.println("Min element: " + peek() + "; H = " + H);

        add(76);
        System.out.println("Min element: " + peek() + "; H = " + H);

        add(19);
        System.out.println("Min element: " + peek() + "; H = " + H);

        add(55);
        System.out.println("Min element: " + peek() + "; H = " + H);

        add(88);
        System.out.println("Min element: " + peek() + "; H = " + H);

        add(2);
        System.out.println("Min element: " + peek() + "; H = " + H);

        System.out.println(H);

        int minValue = extractMin();
        System.out.println("Min element: " + minValue + "; H = " + H);

        minValue = extractMin();
        System.out.println("Min element: " + minValue + "; H = " + H);

        minValue = extractMin();
        System.out.println("Min element: " + minValue + "; H = " + H);

    }

}

File: courses/codes/cclv1-sr-class/heap_example.py
def add(x):
    global H
    H.append(x)
    idx = len(H) - 1
    upheap(idx)

def peek():
    global H
    return H[0]

def extract_min():
    global H
    min_value = peek()
    H = [H[-1]] + H[1:-1]
    downheap(0)
    return min_value


def upheap(i):
    global H
    p = (i - 1) // 2
    if p == -1:
        return
    if H[p] > H[i]:
        H[p], H[i] = H[i], H[p]
        upheap(p)

def downheap(i):
    global H
    lc = 2*i + 1
    rc = 2*i + 2

    if lc >= len(H):
        return
    elif rc >= len(H):
        H[i], H[lc] = H[lc], H[i]
    else:
        if H[lc] <= H[rc]:
            H[i], H[lc] = H[lc], H[i]
            downheap(lc)
        else:
            H[i], H[rc] = H[rc], H[i]
            downheap(rc)


H = []

add(6)
print(f"Min element: {peek()}; H = {H}")

add(50)
print(f"Min element: {peek()}; H = {H}")

add(11)
print(f"Min element: {peek()}; H = {H}")

add(25)
print(f"Min element: {peek()}; H = {H}")

add(42)
print(f"Min element: {peek()}; H = {H}")

add(20)
print(f"Min element: {peek()}; H = {H}")

add(98)
print(f"Min element: {peek()}; H = {H}")

add(76)
print(f"Min element: {peek()}; H = {H}")

add(19)
print(f"Min element: {peek()}; H = {H}")

add(55)
print(f"Min element: {peek()}; H = {H}")

add(88)
print(f"Min element: {peek()}; H = {H}")

add(2)
print(f"Min element: {peek()}; H = {H}")

print(H)

extract_min()
print(f"Min element: {peek()}; H = {H}")

extract_min()
print(f"Min element: {peek()}; H = {H}")

extract_min()
print(f"Min element: {peek()}; H = {H}")